File: Advanced/AutoCompletePractice.py
# Practice 15. Autocomplete system
BUILD_TRIE = 'T'
AUTOCOMPLETE = 'A'
ENDS_HERE = '#'

class Trie:
  def __init__(self):
    self.root = {}
    self.l = []
  def reset_l(self):
    self.l = []

  def build(self, ss):
    for s in ss:
      cur = self.root
      for char in s:
        if char not in cur:
          cur[char] = {}
        cur = cur[char]
      cur[ENDS_HERE] = s

  def search(self, cur):
    if ENDS_HERE in cur:
      self.l.append(cur[ENDS_HERE])
    else:
      for c in cur.keys():
        self.search(cur[c])
  
  def check(self, s):
    cur = self.root
    i = 0
    while i < len(s):
      if s[i] in cur:
        cur = cur[s[i]]
        i += 1
      else:
        break
    if i == len(s):
      self.reset_l()
      self.search(cur)
      return " ".join(self.l)
  
  
if __name__ == '__main__':
  with open('input.txt', 'r') as inFile:
    lines = inFile.readlines()
  with open('output.txt', 'w') as outFile:
    i = 0
    t = Trie()
    while i < len(lines):
      words = lines[i].split()
      op = words[0]

      if op == BUILD_TRIE:
        if len(words) != 2:
          raise Exception("BUILD_TRIE: invalid input")
        n = int(words[1])
        strings = []
        while n:
          i += 1
          strings.append(lines[i].strip())
          n -= 1
        # TODO. Build a trie with "strings"
        t.build(strings)

      elif op == AUTOCOMPLETE:
        if len(words) != 2:
          raise Exception("AUTOCOMPLETE: invalid input")
        prefix = words[1]
        # TODO. Run autocomplete with "prefix" and write results into "outFile"
        outFile.write(t.check(prefix))
        outFile.write("\n")

      else:
        raise Exception("Undefined operator")
      i += 1

File: Advanced/AutoCompletePractice.cpp
// Practice 15. Autocomplete System
#include <algorithm>
#include <fstream>
#include <iostream>
#include <string>
#include <sstream>
#include <vector>
#include <unordered_map>
using namespace std;
const char BUILD_TRIE = 'T';
const char AUTOCOMPLETE = 'A';
const int CHAR_SIZE = 128;

class Trie {
public:
  char data;
  vector<string> l;
  string endsHere;
  unordered_map<char, Trie*> child;
  Trie(): data(0), endsHere(""), child() {}
  Trie(char c): data(c), endsHere(""), child() {}
  void insert(string);
  string check(string);
  void search(Trie*);
};

void Trie::insert(string ss) {
  Trie* root = this;
  for (int i=0; i<ss.length(); i++) {
    if (root->child.find(ss[i]) == root->child.end()) {
      root->child[ss[i]] = new Trie(ss[i]);
    }
    root = root->child[ss[i]];
  }
  root->endsHere = ss;
}

void Trie::search(Trie* cur) {
  if (cur->endsHere != "")
    l.push_back(cur->endsHere);
  else {
    for (unordered_map<char, Trie*>::iterator iter=cur->child.begin(); iter!=cur->child.end(); iter++)
      Trie::search(cur->child[iter->first]);
  }
}

string Trie::check(string s) {
  Trie* root = this;
  int i = 0;
  for (; i<s.length(); i++) {
    if (root->child[s[i]])
      root = root->child[s[i]];
    else
      break;
  }
  if (i == s.length()) {
    l.clear();
    Trie::search(root);
  }
  string tmp = "";
  for (vector<string>::iterator iter=l.begin(); iter!=l.end(); iter++)
    tmp += *iter + " ";
  return tmp;
}


int main() {
  ifstream inFile("input.txt");
  ofstream outFile("output.txt");
  string line;
  Trie t;
  while (getline(inFile, line))
  {
    char op = line[0];
    istringstream iss(line.substr(2));
    vector<string> words;
    string prefix;
    int n;
    switch(op)
    {
      case BUILD_TRIE: 
        if (!(iss >> n)) {
          cerr<<"BUILD_TRIE: invalid input"<<endl;
          exit(1);
        }
        for (int i = 0; i < n; ++i) {
          if (!getline(inFile, line)) {
            cerr<<"BUILD_TRIE: invalid input"<<endl;
            exit(1);
          }
          t.insert(line);
          //words.push_back(line);
        }
        // TODO. Bulid a trie with the stings in vector "words"
        break;

      case AUTOCOMPLETE:
        if (!(iss >> prefix)) {
          cerr<<"AUTOCOMPLETE: invalid input"<<endl;
          exit(1);
        }
        // TODO. Run autocomplete with "prefix" and write result into "outFile"
        outFile << t.check(prefix) << endl;
        break;

      default:
        cerr<<"Undefined operator"<<endl;
        exit(1);
    }
  }
  outFile.close();
  inFile.close();
}
